### PR TITLE
Enhance batch job task management by adding default action types

### PR DESCRIPTION
**Pull request by rbhavna**
### Description
Enhance batch job task management by adding default action types

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Zhangxunmt** [comment]
LGTM overall. Please apply spotLess and add UT coverages to pass CI. Also please define createConnectorAction() in the Util class so it can be shared across multiple classes.

File: common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java [hunk 1]
```diff
@@ -36,6 +37,7 @@ public class ConnectorAction implements ToXContentObject, Writeable {
     public static final String ACTION_PRE_PROCESS_FUNCTION = "pre_process_function";
     public static final String ACTION_POST_PROCESS_FUNCTION = "post_process_function";
 
+
```
Comment (Zhangxunmt) [inline common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java:40]: No need to add this empty line.

File: common/src/main/java/org/opensearch/ml/common/connector/Connector.java [hunk 2]
```diff
@@ -65,6 +65,8 @@ public interface Connector extends ToXContentObject, Writeable {
 
     List<ConnectorAction> getActions();
 
+    void setAction(ConnectorAction action);
```
Comment (Zhangxunmt) [inline common/src/main/java/org/opensearch/ml/common/connector/Connector.java:68]: Using addAction as the name is more accurate?

File: common/src/main/java/org/opensearch/ml/common/output/MLPredictionOutput.java [hunk 3]
```diff
@@ -29,9 +30,11 @@ public class MLPredictionOutput extends MLOutput {
     public static final String TASK_ID_FIELD = "task_id";
     public static final String STATUS_FIELD = "status";
     public static final String PREDICTION_RESULT_FIELD = "prediction_result";
+    public static final String REMOTE_JOB_FIELD = "remote_job";
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/output/MLPredictionOutput.java:33]: let's add a comment about this field.

File: plugin/src/main/java/org/opensearch/ml/action/tasks/CancelBatchJobTransportAction.java [hunk 4]
```diff
@@ -245,4 +254,61 @@ private void processTaskResponse(MLTaskResponse taskResponse, ActionListener<MLC
...
+            .method(method)
+            .url(url)
+            .requestBody(requestBody)
+            .headers(headers)
+            .build();
+
+    }
+
+    private String getEndpointType(String url) {
+        if (url.contains("sagemaker"))
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/tasks/CancelBatchJobTransportAction.java:304]: can we make static variables for all these values and then re-use?

File: plugin/src/main/java/org/opensearch/ml/action/tasks/GetTaskTransportAction.java [hunk 5]
```diff
@@ -362,4 +371,62 @@ private boolean matchesPattern(Pattern pattern, String input) {
...
+            .actionType(BATCH_PREDICT_STATUS)
+            .method(method)
+            .url(url)
+            .requestBody(requestBody)
+            .headers(headers)
+            .build();
+
+    }
+
+    private String getEndpointType(String url) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/tasks/GetTaskTransportAction.java:421]: why do we need same method 2 times?

Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/action/tasks/GetTaskTransportAction.java:421]: Yes will be moving the common methods to ConnectorUtils class

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java [hunk 6]
```diff
@@ -313,4 +318,58 @@ public static SdkHttpFullRequest buildSdkRequest(
...
+            case "cohere":
+                url = isCancelAction ? predictEndpoint + "/${parameters.id}/cancel" : predictEndpoint + "/${parameters.id}";
+                method = isCancelAction ? "POST" : "GET";
+                break;
+            case "bedrock":
+                url = isCancelAction
+                    ? predictEndpoint + "/${parameters.processedJobArn}/stop"
+                    : predictEndpoint + "/${parameters.processedJobArn}";
+                method = isCancelAction ? "POST" : "GET";
+                break;
```
Comment (Zhangxunmt) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:359]: We should add the default branch for this switch statement to return null. In the GetTask and CancelTask, if the ConnectorAction is null, throw an exception with meaning logs like "please provide GetTask/CancelTask action in the connector".

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java [hunk 7]
```diff
@@ -61,6 +63,9 @@ public class ConnectorUtils {
     private static final Aws4Signer signer;
     public static final String SKIP_VALIDATE_MISSING_PARAMETERS = "skip_validating_missing_parameters";
 
+    public static final List<String> SUPPORTED_REMOTE_SERVERS_FOR_DEFAULT_ACTION_TYPES = List
+        .of("sagemaker", "openai", "bedrock", "cohere");
+
```
Comment (Zhangxunmt) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:68]: Ideally if a new platform is used but not listed here, CX should still be able to GetTask and CancelTask by manually adding the actions in the connector. But seems this is not the case in this PR?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java (hunk 6 above)
Comment (rbhavna) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:359]: oh sorry I added it but I guess it got missed during refactoring. Let me add it

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java (hunk 7 above)
Comment (rbhavna) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:68]: Yes if the connector already has the actions configured, then they can get/cancel task for any platform. Only if no action is provided, they we perform this check

**Review by Zhangxunmt — APPROVED** [review]

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**